### PR TITLE
Feat/judge

### LINKED ISSUE #16: feat(judge): Docker sandbox code execution service
## Goal
Build the secure code runner. This is the most security-critical component in the project.

## Implementation
- One isolated Docker container spawned per submission
- Python support for MVP
- Hard resource limits enforced at the Docker level

## Security Requirements (all mandatory)
- [ ] `--network=none` (no outbound network)
- [ ] `--read-only` + `--tmpfs /tmp` (read-only FS with writable temp)
- [ ] `--memory=128m` + `--cpus=0.5` (hard resource caps)
- [ ] `--security-opt=no-new-privileges` (no privilege escalation)
- [ ] Hard kill after 10s timeout enforced by JudgeService (not just container timeout)
- [ ] Containers auto-removed after execution (`--rm`)

## Acceptance Criteria
- [ ] `POST /internal/judge` accepts code + test cases (JSON), returns pass/fail + runtime_ms + memory_kb
- [ ] Fork bomb safely killed within timeout
- [ ] Infinite loop safely killed within timeout
- [ ] JudgeService unit tests with mock Docker client

## Module
Gaming – Complete Web-based Game (Major, 2pts) – judge component

### PR DESCRIPTION
**Pull request by leite-tiago**
## What
Implements the core code execution engine for CodeArena using **Judge0**, a robust, production-ready online judge API. This PR introduces the foundational `JudgeService` in the backend and configures the necessary self-hosted Judge0 infrastructure.

## Why
Code execution is a core feature of the platform. Building a secure and scalable execution sandbox from scratch is complex and error-prone. By integrating Judge0 — a battle-tested engine that uses the `isolate` sandboxing technology (similar to HackerRank and LeetCode) — we guarantee secure, isolated, and highly configurable code execution out of the box, supporting multiple languages safely.

## How
- Created `JudgeService.java` utilizing Spring's `RestClient` to dispatch execution requests (`POST /submissions?wait=true`) to the Judge0 API.
- Implemented `JudgeResponse` and `TestCaseResult` DTOs to parse Judge0's execution status (e.g., Accepted, Time Limit Exceeded, Compilation Error) back to our domain model.
- Updated `docker-compose.yml` to include the required Judge0 cluster (`judge0-server`, `judge0-worker`) along with dedicated PostgreSQL and Redis containers.
- Added `judge0.conf` to configure secure default limitations (CPU time limit, memory limit, network disabled).
- Wrote unit tests for `JudgeServiceTest.java` using Spring's `MockRestServiceServer` to validate behavior across all edge cases (passing, failing, runtime errors, timeouts).

## Screenshots / recordings
N/A (Backend and Infrastructure only)

## Checklist
- [x] All acceptance criteria from the linked issue are met
- [x] `docker compose up --build` works successfully
- [x] No console errors in Chrome DevTools
- [x] Tests added or updated where applicable
- [x] README updated if setup, tech stack, or modules are affected

Closes #16

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -34,6 +34,70 @@ services:
     networks:
       - codearena
 
+  judge0-server:
+    image: judge0/judge0:latest-extra-dev
+    container_name: codearena-judge0-server
+    volumes:
+      - ./judge0.conf:/judge0.conf:ro
+    ports:
+      - "2358:2358"
```
[the remark this answers is not in the document]
Comment (tjlsimoes) [reply docker-compose.yml]: @leite-tiago, is it necessary to add an entryway into the judge from the host?

File: docker-compose.yml [hunk 2]
```diff
@@ -34,6 +34,70 @@ services:
...
+    image: redis:7-alpine
+    container_name: codearena-judge0-cache
+    command: [
+      "sh", "-c",
+      'docker-entrypoint.sh --appendonly no --requirepass "$$REDIS_PASSWORD"'
+    ]
+    env_file: judge0.conf
+    restart: unless-stopped
+    healthcheck:
+      test: ["CMD", "redis-cli", "-p", "6379", "-a", "judge0_redis_secret", "ping"]
```
[the remark this answers is not in the document]
Comment (tjlsimoes) [reply docker-compose.yml]: @leite-tiago do you think this change might be worth applying? Perhaps not using the same redis password already used for our own redis container.

This also raises the question:
- We want to use another Postgres and Redis container just for this judge service?

File: judge0.conf [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+REDIS_HOST=judge0-cache
+REDIS_PORT=6379
+REDIS_PASSWORD=judge0_redis_secret
+
+# PostgreSQL Configuration
+POSTGRES_HOST=judge0-db
+POSTGRES_PORT=5432
+POSTGRES_DB=judge0
+POSTGRES_USER=judge0
+POSTGRES_PASSWORD=judge0_db_secret
```
[the remark this answers is not in the document]
Comment (tjlsimoes) [reply judge0.conf]: @leite-tiago shouldn't we handle the postgres and redis password more securely? Like we do with the other containers? I know this adds friction to something that we perhaps would like to be just "plug and play"...

**Review by tjlsimoes — DISMISSED** [review]
@leite-tiago do you think the changes I've commented on (mainly regarding `docker-compose.yaml`) are worth reviewing? 

I also wasn't able to reach `http://localhost:2358` or `http://localhost:2358/docs` with the current setup (with `JUDGE0_URL=http://localhost:2358` on the `.env` file). I expected to be able to do so, after reading [this](https://github.com/judge0/judge0/blob/master/CHANGELOG.md#v1131-2024-04-18). Is this expected?

**Comment by leite-tiago** [comment]
thanks for the review @tjlsimoes, I will work on these issues!

**Comment by leite-tiago** [comment]
@tjlsimoes now everything should be working, I had to give some privilege to judge0 and the redis/postgres passwords needs to be hardcoded. Otherwise it doens't run.

You should be able to access http://localhost:2358 and http://localhost:2358/docs

**Comment by tjlsimoes** [comment]
@leite-tiago, I can confirm that your changes enabled me to access http://localhost:2358/ and http://localhost:2358/docs . Awesome!

However, the hard-coded passwords for Judge0's Redis and Postgres containers still worry me.

On [42 EvalHub](https://www.42evalhub.com/common/fttranscendence), this shows up on the evaluation sheet:

> Are credentials properly secured:
> 
> - .env file exists and is in .gitignore
> - .env.example file is provided
> - No sensitive credentials in the repository
> 
> Verify these security practices are followed. Any credentials, API keys, or sensitive information found in the repository (outside of .env) will result in immediate project failure.

I think keeping a `judge0.conf` file with harcoded credentials can be interpreted as going against the above technical requirement.

I don't think the values need to be hardcoded. We can perhaps try the following to be safe.

- Keep the `judge0.conf` on the `.gitignore` and make adding it one more step of the deployment process - modifying our existing `setup.sh` to handle this new requirement. 
- Remove the hard-coded credentials for our judge0 service from the `docker-compose.yaml` file. Make them environment variables, definable on our `.env` file.
- Make it so the environment variables defined in the `.env` file are used to generate the `judge0.conf` file.

❓ What do you think? If you want I can tackle this `docker-compose.yaml` and `setup.sh` refactoring myself and see if it works.

**Comment by tjlsimoes** [comment]
@leite-tiago,

- I've removed the `judge0.conf` file. 
- Everything is set from `.env`. So entries were added to `.env.example`. 
- I've disabled access to Judge0 from outside the Docker network, `codearena`. This way we manage to not have a [privileged container](https://docs.docker.com/engine/containers/run/#runtime-privilege-and-linux-capabilities) exposed to the internet, with no authentication!
- Judge0 should now only be accessible to the backend container.

<details>
<summary>You can test this with the following commands (this is a dropdown!)</summary>


First, ensure your containers are running (`docker compose up -d`).

### 1. Test Network Isolation (From your Host)
Since we removed the `ports` mapping, your computer (the host) should no longer be able to "see" Judge0. 

Run this on your terminal:
```bash
curl -I http://localhost:2358/system_info
```
*   **Expected Result**: `curl: (7) Failed to connect to localhost port 2358: Connection refused`. This proves the "wall" is working.

---

### 2. Test Authentication (From inside the Backend)
Even though the backend is inside the network, it still needs the token. We can simulate the backend's perspective using `docker exec`.

#### A. Request WITHOUT the token:
```bash
docker compose exec backend wget -S --spider http://judge0-server:2358/system_info
```
*   **Expected Result**: You should see a `401 Unauthorized` error. This proves that even internal services are blocked if they don't have the key.

#### B. Request WITH the token:
(Replace `your-secret-token` with the actual value you put in your `.env` for `JUDGE0_AUTH_TOKEN`)
```bash
docker compose exec backend wget -q -O - --header="X-Auth-Token: your-secret-token" http://judge0-server:2358/system_info
```
*   **Expected Result**: A JSON response containing system information (CPU, RAM, version). This proves that the backend can talk to Judge0 successfully using the shared secret.

</details>

### I still would like to make sure `setup.sh` covers all the configuration needed to deploy the project. But do tell me if everything works in you end!

**Review by stephan-lassaponari — APPROVED** [review]
